File: C++/StronglyConnectedComponentsKosaraju.cpp
#include <iostream>
#include <vector>
#include <map>
#include <cmath>
#include <algorithm>
#include <deque>
#include <queue>
#include <stack>
#include <set>
#include <string>
#include <string.h>
#include <array>
#include <unordered_map>
#include <unordered_set>
#include <iomanip>
using namespace std;
typedef long long ll;
typedef long double ld;
#define pb push_back

const int MaxN = 1e5+10;
vector<int> g[MaxN], reverseg[MaxN];
bool visited[MaxN];
stack<int> s;
vector<vector<int>> SCC;      // stores all the components
int n, m;

void dfsRev(int cur, vector<int>& curarr) {
    visited[cur] = true;
    curarr.pb(cur);
    for (auto i : reverseg[cur]) {
        if (!visited[i]) dfsRev(i, curarr);
    }
}

void fillstack(int cur) {
    visited[cur] = true;
    for (auto i : g[cur]) {
        if (!visited[i]) fillstack(i);
    }
    s.push(cur);
}

void findSCC() {
    for (int i=0; i<n; i++) {
        if (!visited[i]) fillstack(i);
    }
    
    fill(visited, visited+n, false);
    
    while (!s.empty()) {
        int cur = s.top();
        s.pop();
        if (!visited[cur]) {
            vector<int> curarr;
            dfsRev(cur, curarr);
            SCC.pb(curarr);
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    cin >> n >> m;

    for (int i=0; i<m; i++) {
        int a, b;
        cin >> a >> b;
        a--; b--;
        g[a].pb(b);
        reverseg[b].pb(a);
    }

    findSCC();
    
}

File: Java/StronglyConnectedComponentsKosaraju.java
import java.util.*;
import java.io.*;

public class StronglyConnectedComponentsKosaraju {
	
	static ArrayList<Integer>[] g, reverse;
	static boolean[] visited;
	static ArrayDeque<Integer> stack;
	static ArrayList<ArrayList<Integer>> SCC;
					// stores all of the components
	static int n, m;
	
	public static void main(String[] args) throws IOException {
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		//BufferedReader in = new BufferedReader(new FileReader("StronglyConnectedComponentsKosaraju"));

		StringTokenizer st = new StringTokenizer(in.readLine());
		n = Integer.parseInt(st.nextToken()); 	// number of vertices
		m = Integer.parseInt(st.nextToken()); 	// number of edges
		g = new ArrayList[n];
		reverse = new ArrayList[n];
		stack = new ArrayDeque<>();
		SCC = new ArrayList<>();
		
		visited = new boolean[n];
		for (int i = 0; i < n; i++) {
			g[i] = new ArrayList<>();
			reverse[i] = new ArrayList<>();
		}
		
		for (int i = 0; i < m; i++) {
			st = new StringTokenizer(in.readLine());
			int a = Integer.parseInt(st.nextToken()) - 1;
			int b = Integer.parseInt(st.nextToken()) - 1;
			g[a].add(b);
			reverse[b].add(a);
		}
		
		SCC();
	}
	
	public static void SCC() {
		for (int i = 0; i < n; i++) {
			if (!visited[i]) fillStack(i);
		}
		
		Arrays.fill(visited, false);
		
		while (!stack.isEmpty()) {
			int cur = stack.pop();
			if (!visited[cur]) {
				ArrayList<Integer> curArr = new ArrayList<>();
				dfsRev(cur, curArr);
				SCC.add(curArr);
			}
		}
	}
	
	public static void fillStack(int cur) {
		visited[cur] = true;
		for (Integer i : g[cur]) {
			if (!visited[i]) {
				fillStack(i);
			}
		}
		stack.push(cur);
	}
	
	public static void dfsRev(int cur, ArrayList<Integer> curArr) {
		visited[cur] = true;
		curArr.add(cur);
		for (Integer i : reverse[cur]) {
			if (!visited[i]) {
				dfsRev(i, curArr);
			}
		}
	}
}
